Block Styles: Fix block style variation styles for blocks with complex selectors

aaronrobertshaw opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/62122

## What?

Fixes the style generation for instanced block style variations for blocks that use complex selectors. For example:
- Button: `.wp-block-button .wp-block-button__link`
- Image: 
   - `.wp-block-image`
   - `.wp-block-image img, .wp-block-image .wp-block-image__crop-area, .wp-block-image .components-placeholder`

## Why?

Without this fix, variation styles when applied to an instance would only be applied to the block's wrapper. This means styles like the Button block's outline block style would get double borders, padding et.c

## How?

Reworks the generation of the instanced block style variation stylesheets so that the result has selectors that apply different types of styles to the block's desired elements. For example, the Button block styles go to the inner `.wp-block-button__link` element or an Image's border goes to the inner `img`

The style load order was also tweaked to ensure the block style variations styles were loaded after the block library styles. 

## Testing Instructions

1. Register some custom block style variations following [#57908](https://github.com/WordPress/gutenberg/pull/57908)'s instructions
2. Apply these in a nested fashion and ensure they still continue to work
3. Make sure you have a button block in your post or page
4. Apply the outline block style to the button
5. Make sure all the blocks display correctly in both editors and the frontend
6. Disable separate core block assets via snippet below or activate a theme which does such as [Powder](https://github.com/bgardner/powder/tree/main)
7. Confirm block library styles do not override the outline block style variation styles for Button blocks

`add_filter( 'should_load_separate_core_block_assets', '__return_false' );`

## Screenshots or screencast <!-- if applicable -->

| Before | After |
|---|---|
| <img width="1416" alt="Screenshot 2024-05-30 at 4 25 25 PM" src="https://github.com/WordPress/gutenberg/assets/60436221/1b75d076-fee0-44f9-92d2-8b736ae70069"> | <img width="1375" alt="Screenshot 2024-05-30 at 4 25 08 PM" src="https://github.com/WordPress/gutenberg/assets/60436221/f63af947-7fda-4f7c-af9a-42e52b959df3"> |

aaronrobertshaw committed:
Block Styles: Fix block style variation styles for blocks with complex selectors

aaronrobertshaw committed:
Add backport log

aaronrobertshaw committed:
Fix multiline comment format

aaronrobertshaw commented:
The changes here have been included within the overall backport for the section styles feature: https://github.com/WordPress/wordpress-develop/pull/6662.

If this PR requires further updates, that backport will also need updating.

talldan commented on lib/block-supports/block-style-variations.php:
```diff
@@ -79,31 +79,63 @@ function gutenberg_render_block_style_variation_support_styles( $parsed_block )
-
 	$class_name         = gutenberg_get_block_style_variation_class_name( $parsed_block, $variation );
 	$updated_class_name = $parsed_block['attrs']['className'] . " $class_name";
+	$variation_instance = substr( $class_name, 9 );
```
It feels unusual now that the class name is created, to immediately remove part of it here.

Maybe `gutenberg_get_block_style_variation_class_name` could be changed to `gutenberg_get_block_style_variation_instance`, and then the `is-style-` part appended on after (there could still be a `gutenberg_get_block_style_variation_class_name` function if needed for that but it accepts the variation_instance).

aaronrobertshaw replied:
Done https://github.com/WordPress/gutenberg/pull/62125/commits/a4662f2256a08331ec8685e919e1b97297910a67

I'll update the backport accordingly shortly.

talldan commented on lib/block-supports/block-style-variations.php:
```diff
@@ -79,31 +79,63 @@ function gutenberg_render_block_style_variation_support_styles( $parsed_block )
+	$class_name         = ".$class_name";
+
+	/*
+	 * To support blocks with more complex selectors, that can apply styles to
+	 * inner markup or even different inner elements depending on the feature,
+	 * the variation style data needs to be manipulated some. If the root variation
+	 * styles are moved under a variations property they will generate as desired.
```
I'm not quite following this TBH, might need to explain it more. 😄

aaronrobertshaw replied:
I was trying to avoid writing an essay here. Would linking to the issue and PR help?

aaronrobertshaw replied:
I have a better idea on how best to explain this bit of the code now. I'll update the inline comment shortly.

For the discussion on this PR though:

The crux of the issue is that blocks can define their own selectors to direct certain styles to inner parts of the blocks markup. For example, the inner `img` element for an Image block needs to get the border styles whereas the other styles might still belong on the outer `figure` wrapper for the block.

Previously, the variation's data was treated as its own "theme.json partial". The problem here is that root styles for theme.json don't have the ability to be directed to inner parts as that's what the `styles.elements` and `styles.blocks` properties do.

The fix in this PR essentially moves the variation's root styles to under the `styles.blocks` property so that the block's custom selectors can be honored.

aaronrobertshaw replied:
Done https://github.com/WordPress/gutenberg/pull/62125/commits/2f33a3d130d8a84f4659c6f94e88abc2f9063430

talldan reviewed (commented):
It does fix the issue, but I don't understand the fix, so I'd have a hard time approving it.

I do notice there are two sets of identical styles:
![Screenshot 2024-05-30 at 4 31 04 pm](https://github.com/WordPress/gutenberg/assets/677833/2b0bef50-d373-4a9d-aafe-3fc02e4b15e1)

If there's any way to avoid that it might be good, but it could be considered a separate change.

aaronrobertshaw committed:
Address feedback

aaronrobertshaw committed:
Refactor class name generation as suggested

aaronrobertshaw commented:
Appreciate the thorough review @talldan 🙇 

I believe I've addressed all the feedback so far. Should be good to go for another spin now.

> If there's any way to avoid that it might be good, but it could be considered a separate change.

I think that is definitely best done via a separate follow-up. 

For the record, one set of styles is from the standard theme.json stylesheet but can't be relied upon as if variations are nested the load order needs to be different for the correct variation's styles to take precedence. That's where the second set of styles comes in as they are the styles specific to this individual application or instance of the block style variation.

The code creating the standard set of styles is actually leveraged to generate the individual instance styles. 

It is 100% something that would be good to clean up. Right now though I think we need to land a fix for the issue affecting outline buttons etc.

aaronrobertshaw committed:
Fix missed reference to old function name

bgardner commented:
@aaronrobertshaw 

There seems to still be an issue with specificity and styles. My Powder theme which uses theme.json to style the outline variation (https://github.com/bgardner/powder/blob/main/theme.json#L288) is being overridden by the following:

```
:root :where(.wp-block-button .wp-block-button__link.is-style-outline), :root :where(.wp-block-button.is-style-outline>.wp-block-button__link) {
    border: 2px solid;
    padding: .667em 1.333em;
}
```

aaronrobertshaw commented:
Thanks for the extra information @bgardner, it definitely helps 👍 

After downloading your theme and poking around a bit, it looks like the style load order isn't quite right. The key difference I noted while testing was that Powder disables the separate block assets via the `should_load_separate_core_block_assets` filter.

The `wp-block-library-inline-css` styles are added after the global styles and block style variation styles. When the block styles are separated however their individual stylesheets are loaded before the global styles and variation styles.

I'll work on a fix for this first thing tomorrow. It might also be better as a separate PR.

bgardner commented:
@aaronrobertshaw That makes sense, thanks for spelling that out. If there's an alternative (more optimal) way for me to handle things in Powder, I'm all ears.

aaronrobertshaw commented:
> If there's an alternative (more optimal) way for me to handle things in Powder, I'm all ears.

I don't think you should need to make any changes here. Hopefully, a bit of tweaking to the order styles load in should see this working as expected.

aaronrobertshaw committed:
Ensure the variation styles are loaded after block library styles

aaronrobertshaw commented:
@bgardner with fresh eyes this morning it looks like the tweak the load order is rather trivial so I've rolled it into this fix.

| Before | After |
|---|---|
| <img width="262" alt="Screenshot 2024-05-31 at 9 55 29 AM" src="https://github.com/WordPress/gutenberg/assets/60436221/2aaa8b7a-b28f-4d67-86eb-0871eb6f5836"> | <img width="269" alt="Screenshot 2024-05-31 at 9 55 09 AM" src="https://github.com/WordPress/gutenberg/assets/60436221/4aff1245-f614-4d40-a1f1-8d4533c35c22"> |

bgardner commented:
@aaronrobertshaw Can confirm, per the below. LGTM!

![image](https://github.com/WordPress/gutenberg/assets/486261/b9ec8170-482f-4a89-beb1-e66e75af0286)

talldan reviewed (approved):
Looks good now, thanks for adding the improved explanations in the code, that helped me understand the issue and fix.

aaronrobertshaw commented:
Appreciate all the reviews and testing. Thanks @talldan and @bgardner 🙇
